core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -453,7 +456,15 @@ public void onFailure(String source, @Nullable Throwable t) {
      *
      * @return true if successful
      */
-    private boolean joinElectedMaster(DiscoveryNode masterNode) {
+    boolean joinElectedMaster(DiscoveryNode masterNode) {
+        // Sanity checks: likely we don't end up here, because serialization is likely to fail.
+        if (masterNode.getVersion().before(version.minimumCompatibilityVersion())) {
```

Comment:
I think this can be better put in the findmaster() logic. If the "elected" master is not of a good version we should log a warning and return null (no master found)